stash_api/app/controllers/stash_api/application_controller.rb

```diff
@@ -77,10 +77,15 @@ def require_in_progress_resource
       @resource = @stash_identifier.in_progress_resource
     end
 
+    def require_viewable_resource(resource_id:)
+      res = StashEngine::Resource.where(id: resource_id).first
+      render json: { error: 'not-found' }.to_json, status: 404 if res.nil? || !res.may_view?(ui_user: @user)
```

Comment:
do you want to distinguish between the resource not being found and a permission issue? Maybe:
```ruby
render json: { error: 'not-found' }.to_json, status: 404 if res.nil? 
render json: { error: 'unauthorized' }.to_json, status: 401 unless res.may_view?(ui_user: @user)
```